thirdparty/hnswlib/hnswlib/hnswalg.h

```diff
@@ -1424,15 +1452,18 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
         auto accumulative_alpha = (bitset.count() >= (cur_element_count * kHnswSearchKnnBFFilterThreshold))
                                       ? std::numeric_limits<float>::max()
                                       : 0.0f;
-        if (metric_type_ == Metric::COSINE) {
-            return std::make_unique<IteratorWorkspace>(
-                query_data, max_elements_, seed_ef, for_tuning,
-                knowhere::CopyAndNormalizeVecs((const float*)query_data, 1, *(size_t*)dist_func_param_), bitset,
-                accumulative_alpha);
-        } else {
-            return std::make_unique<IteratorWorkspace>(query_data, max_elements_, seed_ef, for_tuning, nullptr, bitset,
-                                                       accumulative_alpha);
+        if constexpr (knowhere::KnowhereFloatTypeCheck<data_t>::value) {
+            if (metric_type_ == Metric::COSINE) {
+                auto normalized_query = std::make_unique<char[]>(space_->get_data_size());
+                std::memcpy(normalized_query.get(), query_data, space_->get_data_size());
```

Comment:
memcpy_s maybe  safer?

Reply:
`  memcpy_s(
     void *dest,
     size_t numberOfElements,
     const void *src,
     size_t count 
  );` numberOfElements and  count cannot be distinguished in these code, and memcpy_s is not as portable as memcpy.